### DISCUSSION TITLE
EI thread

### DISCUSSION THREAD

**Discussion by OOHehir** [LIFsCode/ELOC-3.0#80] [General]
@LIFsCode This discussion is to pick up where we left off about terminating the EI thread once inference is halted.

Currently, the I2S thread runs continuously as soon as the recording or inference is started. Do you think it'll be required functionality at some point that the ELOC needs to hibernate/deep sleep during certain hours & then resume? If so terminating & then restarting the EdgeImpulse thread might be safer. If this functionality won't be required, I can make the thread non-terminating.

**Comment by LIFsCode** [comment]
@OOHehir I will draw a few diagrams to depict the potential operation modes of the ELOC at least as far as I've understood it.

But yes I guess a temporary disabling/stopping if recording, detection and maybe even the I2S sampling may be relevant.

But what do you mean with "may be safer" if the thread is terminated? I get the point that it's more efficient from a memory view and has a minimal performance impact, but in what way safer?

Actually I would suppose it's safer to leave it alive but just suspend it (requires setting [INCLUDE_vTaskSuspend](https://www.freertos.org/a00118.html)
Thus all handles remain valid, no danger of heap fragmentation, no danger of invalid access.

**Reply by OOHehir** [reply]
@LIFsCode My concern is that the ELOC is restarted with a change in sample rate (due to change in setting by user). This would need to be handled correctly by the inference (& wav writer). It shouldn't be an issue but just needs some added functionality. 

Most of the heap use/ churn probably results from the inference process, I've tried to minimize this by the build flag [EI_CLASSIFIER_ALLOCATION_STATIC](https://docs.edgeimpulse.com/docs/run-inference/cpp-library#static-allocation) but some is unavoidable.
